### PR TITLE
feat: Allow sharing to and from team projects (no-changelog)

### PR DESCRIPTION
**Pull request by valya**
## Summary

Remove the backend restrictions on sharing to and from team projects. Also, make it so that transferring a credential won't sever all shares.

## Related Linear tickets, Github issues, and Community forum posts

[PAY-1741](https://linear.app/n8n/issue/PAY-1741/enable-sharing-of-credentials-between-personal-and-team-projects)

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) 
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included. 
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

### CODE REVIEW

File: packages/cli/test/integration/credentials/credentials.api.ee.test.ts [hunk 1]
```diff
@@ -978,6 +980,128 @@ describe('PUT /credentials/:id/share', () => {
 
 		config.set('userManagement.emails.mode', 'smtp');
 	});
+
+	test('should be able to share from personal project to team project that member has access to', async () => {
```
Comment (ivov) [inline packages/cli/test/integration/credentials/credentials.api.ee.test.ts:984]:
```suggestion
	test('member should be able to share from personal project to team project that member has access to', async () => {
```

File: packages/cli/test/integration/credentials/credentials.api.ee.test.ts [hunk 2]
```diff
@@ -978,6 +980,128 @@ describe('PUT /credentials/:id/share', () => {
...
+		expect(response.statusCode).toBe(200);
+		expect(response.body.data).toBeUndefined();
+
+		const shares = await getCredentialSharings(savedCredential);
+		const testShare = shares.find((s) => s.projectId === testProject.id);
+		expect(testShare).not.toBeUndefined();
+		expect(testShare?.role).toBe('credential:user');
+	});
+
+	test('should be able to share from team project to personal project', async () => {
```
Comment (ivov) [inline packages/cli/test/integration/credentials/credentials.api.ee.test.ts:1003]: I'd prefer to add roles here to better keep track of what we're testing in each case. 

```suggestion
	test('member should be able to share from team project to personal project', async () => {
```

File: packages/cli/test/integration/credentials/credentials.api.ee.test.ts [hunk 3]
```diff
@@ -978,6 +980,128 @@ describe('PUT /credentials/:id/share', () => {
...
+		expect(response.statusCode).toBe(200);
+		expect(response.body.data).toBeUndefined();
+
+		const shares = await getCredentialSharings(savedCredential);
+		const testShare = shares.find((s) => s.projectId === anotherMemberPersonalProject.id);
+		expect(testShare).not.toBeUndefined();
+		expect(testShare?.role).toBe('credential:user');
+	});
+
+	test('should be able to share from team project to team project that member has access to', async () => {
```
Comment (ivov) [inline packages/cli/test/integration/credentials/credentials.api.ee.test.ts:1023]:
```suggestion
	test('member should be able to share from team project to team project where member has access to both', async () => {
```

File: packages/cli/test/integration/credentials/credentials.api.ee.test.ts [hunk 4]
```diff
@@ -978,6 +980,128 @@ describe('PUT /credentials/:id/share', () => {
...
+
+	test('admins should be able to share from any team project to team project ', async () => {
+		const testProject = await createTeamProject();
+		const testProject2 = await createTeamProject();
+
+		const savedCredential = await saveCredential(randomCredentialPayload(), {
+			project: testProject,
+		});
+
+		const response = await authOwnerAgent
```
Comment (ivov) [inline packages/cli/test/integration/credentials/credentials.api.ee.test.ts:1053]: A bit confused - test title says admins but here we're using the owner agent? Shouldn't we be using an `authAdminAgent`?

File: packages/cli/test/integration/credentials/credentials.api.ee.test.ts [hunk 5]
```diff
@@ -978,6 +980,128 @@ describe('PUT /credentials/:id/share', () => {
...
+		expect(response.statusCode).toBe(200);
+		expect(response.body.data).toBeUndefined();
+
+		const shares = await getCredentialSharings(savedCredential);
+		const testShare = shares.find((s) => s.projectId === testProject2.id);
+		expect(testShare).not.toBeUndefined();
+		expect(testShare?.role).toBe('credential:user');
+	});
+
+	test('admins should be able to share from any team project to personal project ', async () => {
```
Comment (ivov) [inline packages/cli/test/integration/credentials/credentials.api.ee.test.ts:1066]: To emphasize that the destination project doesn't need to be the admin's personal project.

```suggestion
	test('admins should be able to share from any team project to any user\'s personal project', async () => {
```

File: packages/cli/test/integration/credentials/credentials.api.ee.test.ts [hunk 6]
```diff
@@ -978,6 +980,128 @@ describe('PUT /credentials/:id/share', () => {
...
+		expect(response.statusCode).toBe(200);
+		expect(response.body.data).toBeUndefined();
+
+		const shares = await getCredentialSharings(savedCredential);
+		const testShare = shares.find((s) => s.projectId === memberPersonalProject.id);
+		expect(testShare).not.toBeUndefined();
+		expect(testShare?.role).toBe('credential:user');
+	});
+
+	test('admins should be able to share from any personal project to team project ', async () => {
```
Comment (ivov) [inline packages/cli/test/integration/credentials/credentials.api.ee.test.ts:1086]:
```suggestion
	test('admins should be able to share from any personal project to any team project', async () => {
```

File: packages/cli/test/integration/credentials/credentials.api.ee.test.ts [hunk 7]
```diff
@@ -1084,12 +1208,72 @@ describe('PUT /:credentialId/transfer', () => {
...
+				expect.objectContaining({
+					credentialsId: credential.id,
+					projectId,
+					role: 'credential:user',
+				}),
+			);
+		}
+	});
+
+	test('transferring should share with original owner when asked', async () => {
```
Comment (ivov) [inline packages/cli/test/integration/credentials/credentials.api.ee.test.ts:1230]:
```suggestion
	test('transferring should share with original credential owner's personal project when asked', async () => {
```

**Review by ivov — DISMISSED** [review]
Far from familiar with RBAC but the logic makes sense to me.

Only renamings and questions to better understand, feel free to disregard.

**Review by ivov — DISMISSED** [review]

File: packages/cli/test/integration/credentials/credentials.api.ee.test.ts (hunk 4 above)
Comment (ivov) [reply packages/cli/test/integration/credentials/credentials.api.ee.test.ts:1053]: Got you. We need a more accurate umbrella term but can't of one right now. `superuser` might not be immediately clear. Let's move on for now.

**Review by ivov — DISMISSED** [review]
Thanks for addressing everything! Sorry for the delay.

**Review by ivov — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@1.55.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.55.0)

**Pull request merged**